File: Tangra 3/Video/AstroDigitalVideo/AdvFramesIndex.cs
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

namespace Tangra.Video.AstroDigitalVideo
{
    public class AdvFramesIndexEntry
    {
        public TimeSpan ElapsedTime;
        public long Offset;
        public uint Length;
    }

    public class AdvFramesIndex
    {
        private List<AdvFramesIndexEntry> m_Index = new List<AdvFramesIndexEntry>();

        public uint NumberOfFrames { get { return (uint)m_Index.Count; } }

		public long TableOffset { get; private set; }

	    public IList<AdvFramesIndexEntry> Index
        {
            get { return m_Index; }
        }

        public AdvFramesIndex()
        { }

        private void AddFrame(long elapsedMilliseconds, long frameOffset, uint size)
        {
            m_Index.Add(new AdvFramesIndexEntry()
            {
                ElapsedTime = TimeSpan.FromMilliseconds(elapsedMilliseconds),
                Offset = frameOffset,
                Length = size
            });
        }

        public AdvFramesIndex(BinaryReader reader)
        {
	        TableOffset = reader.BaseStream.Position;

            uint count = reader.ReadUInt32();

            for (int i = 0; i < count; i++)
            {
				uint elapsedMiliseconds = reader.ReadUInt32();
                long offset = reader.ReadInt64();
                uint length = reader.ReadUInt32();                

                AddFrame(elapsedMiliseconds, offset, length);
            }
        }
    }
}

File: AdvLib.Core/src/adv_frames_index.cpp
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

#include "StdAfx.h"
#include "adv_frames_index.h"
#include "stdio.h"

namespace AdvLib
{
	
AdvFramesIndex::AdvFramesIndex()
{
	m_IndexEntries = new vector<IndexEntry*>();
}

AdvFramesIndex::~AdvFramesIndex()
{
	m_IndexEntries->clear();
	delete m_IndexEntries;
}

void AdvFramesIndex::AddFrame(unsigned int frameNo, unsigned int elapedTime, __int64 frameOffset, unsigned int  bytesCount)
{
	IndexEntry *entry = new IndexEntry();
	entry->BytesCount = bytesCount;
	entry->FrameOffset = frameOffset;
	entry->ElapsedTime = elapedTime;
	
	m_IndexEntries->push_back(entry);
}

void AdvFramesIndex::WriteIndex(FILE *pFile)
{
	unsigned int framesCount = m_IndexEntries->size();
	fwrite(&framesCount, 4, 1, pFile);

	vector<IndexEntry*>::iterator curr = m_IndexEntries->begin();
	while (curr != m_IndexEntries->end()) 
	{
		unsigned int elapedTime = (*curr)->ElapsedTime;
		__int64 frameOffset = (*curr)->FrameOffset;
		unsigned int  bytesCount = (*curr)->BytesCount;
		
		fwrite(&elapedTime, 4, 1, pFile);
		fwrite(&frameOffset, 8, 1, pFile);
		fwrite(&bytesCount, 4, 1, pFile);
		
		curr++;
	}
}

}

